File: node/src/guardrails/streaming.ts
/**
 * Streaming guardrails for real-time token-by-token LLM output validation.
 *
 * This module provides the StreamingGuardrail class for evaluating LLM output
 * tokens as they are generated, enabling early termination on policy violations.
 *
 * @example
 * ```typescript
 * import { StreamingGuardrail } from '@diagnyx/node';
 *
 * const guardrail = new StreamingGuardrail({
 *   apiKey: 'dx_...',
 *   organizationId: 'org_123',
 *   projectId: 'proj_456',
 * });
 *
 * // Start a session
 * const session = await guardrail.startSession();
 *
 * // Evaluate tokens as they arrive
 * for await (const chunk of openaiStream) {
 *   const token = chunk.choices[0]?.delta?.content || '';
 *   const isLast = chunk.choices[0]?.finish_reason !== null;
 *
 *   for await (const filteredToken of guardrail.evaluate(token, { isLast })) {
 *     process.stdout.write(filteredToken);
 *   }
 * }
 * ```
 */

import { io, Socket } from 'socket.io-client';

/**
 * Enforcement level for guardrail policies
 */
export type EnforcementLevel = 'advisory' | 'warning' | 'blocking';

/**
 * Types of streaming events
 */
export type StreamingEventType =
  | 'session_started'
  | 'token_allowed'
  | 'violation_detected'
  | 'early_termination'
  | 'session_complete'
  | 'error'
  | 'heartbeat';

/**
 * Configuration for StreamingGuardrail
 */
export interface StreamingGuardrailConfig {
  apiKey: string;
  organizationId: string;
  projectId: string;
  baseUrl?: string;
  wsUrl?: string;
  timeout?: number;
  evaluateEveryNTokens?: number;
  enableEarlyTermination?: boolean;
  useWebSocket?: boolean;
  debug?: boolean;
}

/**
 * Guardrail violation details
 */
export interface GuardrailViolation {
  policyId: string;
  policyName: string;
  policyType: string;
  violationType: string;
  message: string;
  severity: string;
  enforcementLevel: EnforcementLevel;
  details?: Record<string, unknown>;
}

/**
 * Streaming session state
 */
export interface StreamingSession {
  sessionId: string;
  organizationId: string;
  projectId: string;
  activePolicies: string[];
  tokensProcessed: number;
  violations: GuardrailViolation[];
  terminated: boolean;
  terminationReason?: string;
  allowed: boolean;
  accumulatedText: string;
}

/**
 * Options for starting a session
 */
export interface StartSessionOptions {
  input?: string;
}

/**
 * Options for evaluating a token
 */
export interface EvaluateOptions {
  tokenIndex?: number;
  isLast?: boolean;
}

/**
 * Error thrown when a blocking guardrail violation terminates the stream
 */
export class GuardrailViolationError extends Error {
  constructor(
    public readonly violation: GuardrailViolation,
    public readonly session: StreamingSession,
  ) {
    super(`Guardrail violation: ${violation.message}`);
    this.name = 'GuardrailViolationError';
  }
}

/**
 * Token-by-token streaming guardrail for LLM output validation.
 *
 * Provides real-time evaluation of LLM response tokens against configured
 * guardrail policies with support for early termination on blocking violations.
 *
 * Supports both HTTP SSE and WebSocket connections for streaming evaluation.
 */
export class StreamingGuardrail {
  private readonly config: Required<StreamingGuardrailConfig>;
  private session: StreamingSession | null = null;
  private tokenIndex = 0;
  private wsClient: Socket | null = null;

  constructor(config: StreamingGuardrailConfig) {
    const baseUrl = config.baseUrl?.replace(/\/$/, '') || 'https://api.diagnyx.io';
    const wsUrl =
      config.wsUrl ||
      (baseUrl.startsWith('https://')
        ? baseUrl.replace('https://', 'wss://') + '/guardrails'
        : baseUrl.replace('http://', 'ws://') + '/guardrails');

    this.config = {
      apiKey: config.apiKey,
      organizationId: config.organizationId,
      projectId: config.projectId,
      baseUrl,
      wsUrl,
      timeout: config.timeout || 30000,
      evaluateEveryNTokens: config.evaluateEveryNTokens || 10,
      enableEarlyTermination: config.enableEarlyTermination ?? true,
      useWebSocket: config.useWebSocket || false,
      debug: config.debug || false,
    };
  }

  private log(message: string): void {
    if (this.config.debug) {
      console.log(`[DiagnyxGuardrails] ${message}`);
    }
  }

  private getHeaders(): Record<string, string> {
    return {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${this.config.apiKey}`,
      Accept: 'text/event-stream',
    };
  }

  private getBaseEndpoint(): string {
    return `${this.config.baseUrl}/api/v1/organizations/${this.config.organizationId}/guardrails`;
  }

  /**
   * Start a new streaming guardrail session
   */
  async startSession(options: StartSessionOptions = {}): Promise<StreamingSession> {
    if (this.config.useWebSocket) {
      return this.startWsSession(options);
    }

    const payload = {
      projectId: this.config.projectId,
      evaluateEveryNTokens: this.config.evaluateEveryNTokens,
      enableEarlyTermination: this.config.enableEarlyTermination,
      ...(options.input && { input: options.input }),
    };

    const response = await fetch(`${this.getBaseEndpoint()}/evaluate/stream/start`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new Error(`Failed to start session: ${response.statusText}`);
    }

    const data = await response.json();

    if (data.type === 'session_started') {
      this.session = {
        sessionId: data.sessionId,
        organizationId: this.config.organizationId,
        projectId: this.config.projectId,
        activePolicies: data.activePolicies || [],
        tokensProcessed: 0,
        violations: [],
        terminated: false,
        allowed: true,
        accumulatedText: '',
      };
      this.tokenIndex = 0;
      this.log(`Session started: ${this.session.sessionId}`);
      return this.session;
    } else if (data.type === 'error') {
      throw new Error(`Failed to start session: ${data.error}`);
    }

    throw new Error(`Unexpected response type: ${data.type}`);
  }

  /**
   * Start session via WebSocket
   */
  private async startWsSession(options: StartSessionOptions): Promise<StreamingSession> {
    return new Promise((resolve, reject) => {
      this.wsClient = io(this.config.wsUrl, {
        auth: { token: this.config.apiKey },
        transports: ['websocket'],
      });

      this.wsClient.on('connect', () => {
        this.log('WebSocket connected');
        this.wsClient?.emit('start_session', {
          projectId: this.config.projectId,
          input: options.input,
          evaluateEveryNTokens: this.config.evaluateEveryNTokens,
          enableEarlyTermination: this.config.enableEarlyTermination,
        });
      });

      this.wsClient.on('session_started', (data: any) => {
        this.session = {
          sessionId: data.sessionId,
          organizationId: this.config.organizationId,
          projectId: this.config.projectId,
          activePolicies: data.activePolicies || [],
          tokensProcessed: 0,
          violations: [],
          terminated: false,
          allowed: true,
          accumulatedText: '',
        };
        this.tokenIndex = 0;
        this.log(`Session started (WS): ${this.session.sessionId}`);
        resolve(this.session);
      });

      this.wsClient.on('error', (error: any) => {
        reject(new Error(`WebSocket error: ${error.error || error}`));
      });

      this.wsClient.on('connect_error', (error: Error) => {
        reject(new Error(`WebSocket connection failed: ${error.message}`));
      });
    });
  }

  /**
   * Evaluate a token against guardrail policies
   *
   * Yields the token if it passes validation. Throws GuardrailViolationError
   * if a blocking violation is detected and early termination is enabled.
   */
  async *evaluate(token: string, options: EvaluateOptions = {}): AsyncGenerator<string> {
    if (!this.session) {
      throw new Error('No active session. Call startSession() first.');
    }

    if (this.config.useWebSocket && this.wsClient) {
      yield* this.evaluateWs(token, options);
      return;
    }

    const payload = {
      sessionId: this.session.sessionId,
      token,
      tokenIndex: options.tokenIndex ?? this.tokenIndex,
      isLast: options.isLast ?? false,
    };

    this.session.accumulatedText += token;
    this.tokenIndex++;

    const response = await fetch(`${this.getBaseEndpoint()}/evaluate/stream`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new Error(`Token evaluation failed: ${response.statusText}`);
    }

    const reader = response.body?.getReader();
    if (!reader) {
      throw new Error('No response body');
    }

    const decoder = new TextDecoder();
    let buffer = '';

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() || '';

        for (const line of lines) {
          if (!line.startsWith('data: ')) continue;

          try {
            const data = JSON.parse(line.slice(6));
            const eventType = data.type;

            if (eventType === 'token_allowed') {
              this.session.tokensProcessed = (data.tokenIndex ?? 0) + 1;
              yield token;
            } else if (eventType === 'violation_detected') {
              const violation = this.parseViolation(data);
              this.session.violations.push(violation);

              if (violation.enforcementLevel === 'blocking') {
                this.session.allowed = false;
              }
            } else if (eventType === 'early_termination') {
              const violation = this.parseViolation(data.blockingViolation || {});
              this.session.terminated = true;
              this.session.terminationReason = data.reason;
              this.session.allowed = false;
              throw new GuardrailViolationError(violation, this.session);
            } else if (eventType === 'session_complete') {
              this.session.tokensProcessed = data.totalTokens ?? 0;
              this.session.allowed = data.allowed ?? true;
            } else if (eventType === 'error') {
              this.log(`Error: ${data.error}`);
            }
          } catch (e) {
            if (e instanceof GuardrailViolationError) throw e;
            this.log(`Failed to parse event: ${e}`);
          }
        }
      }
    } finally {
      reader.releaseLock();
    }
  }

  /**
   * Evaluate token via WebSocket
   */
  private async *evaluateWs(token: string, options: EvaluateOptions): AsyncGenerator<string> {
    if (!this.wsClient || !this.session) {
      throw new Error('WebSocket not connected');
    }

    const session = this.session;
    session.accumulatedText += token;
    const currentIndex = this.tokenIndex++;

    return yield* new Promise<AsyncGenerator<string>>((resolve) => {
      const results: string[] = [];
      let resolved = false;

      const handlers = {
        token_allowed: (data: any) => {
          session.tokensProcessed = (data.tokenIndex ?? 0) + 1;
          results.push(token);
          cleanup();
          resolve(
            (async function* () {
              yield token;
            })(),
          );
        },
        violation_detected: (data: any) => {
          const violation = this.parseViolation(data);
          session.violations.push(violation);
          if (violation.enforcementLevel === 'blocking') {
            session.allowed = false;
          }
        },
        early_termination: (data: any) => {
          const violation = this.parseViolation(data.blockingViolation || {});
          session.terminated = true;
          session.terminationReason = data.reason;
          session.allowed = false;
          cleanup();
          resolve(
            (async function* () {
              throw new GuardrailViolationError(violation, session);
            })(),
          );
        },
        session_complete: (data: any) => {
          session.tokensProcessed = data.totalTokens ?? 0;
          session.allowed = data.allowed ?? true;
          cleanup();
          resolve(
            (async function* () {
              // Nothing to yield on complete
            })(),
          );
        },
      };

      const cleanup = () => {
        if (resolved) return;
        resolved = true;
        Object.keys(handlers).forEach((event) => {
          this.wsClient?.off(event, (handlers as any)[event]);
        });
      };

      Object.keys(handlers).forEach((event) => {
        this.wsClient?.on(event, (handlers as any)[event]);
      });

      this.wsClient?.emit('evaluate_token', {
        sessionId: session.sessionId,
        token,
        tokenIndex: options.tokenIndex ?? currentIndex,
        isLast: options.isLast ?? false,
      });
    });
  }

  /**
   * Complete the current session
   */
  async completeSession(): Promise<StreamingSession> {
    if (!this.session) {
      throw new Error('No active session');
    }

    if (this.config.useWebSocket && this.wsClient) {
      return this.completeWsSession();
    }

    const response = await fetch(
      `${this.getBaseEndpoint()}/evaluate/stream/${this.session.sessionId}/complete`,
      {
        method: 'POST',
        headers: this.getHeaders(),
      },
    );

    if (!response.ok) {
      throw new Error(`Failed to complete session: ${response.statusText}`);
    }

    const reader = response.body?.getReader();
    if (reader) {
      const decoder = new TextDecoder();
      let buffer = '';

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() || '';

        for (const line of lines) {
          if (!line.startsWith('data: ')) continue;
          try {
            const data = JSON.parse(line.slice(6));
            if (data.type === 'session_complete') {
              this.session.tokensProcessed = data.totalTokens ?? 0;
              this.session.allowed = data.allowed ?? true;
            }
          } catch {
            // Skip invalid JSON
          }
        }
      }
      reader.releaseLock();
    }

    const session = this.session;
    this.session = null;
    return session;
  }

  /**
   * Complete session via WebSocket
   */
  private async completeWsSession(): Promise<StreamingSession> {
    return new Promise((resolve) => {
      if (!this.wsClient || !this.session) {
        throw new Error('WebSocket not connected');
      }

      const session = this.session;

      this.wsClient.once('session_complete', (data: any) => {
        session.tokensProcessed = data.totalTokens ?? 0;
        session.allowed = data.allowed ?? true;
        this.session = null;
        resolve(session);
      });

      this.wsClient.emit('complete_session', {
        sessionId: session.sessionId,
      });
    });
  }

  /**
   * Cancel the current session
   */
  async cancelSession(): Promise<boolean> {
    if (!this.session) {
      return false;
    }

    if (this.config.useWebSocket && this.wsClient) {
      this.wsClient.emit('cancel_session', {
        sessionId: this.session.sessionId,
      });
      this.session = null;
      return true;
    }

    const response = await fetch(
      `${this.getBaseEndpoint()}/evaluate/stream/${this.session.sessionId}`,
      {
        method: 'DELETE',
        headers: this.getHeaders(),
      },
    );

    if (!response.ok) {
      throw new Error(`Failed to cancel session: ${response.statusText}`);
    }

    const data = await response.json();
    this.session = null;
    return data.cancelled ?? false;
  }

  /**
   * Get the current session
   */
  getSession(): StreamingSession | null {
    return this.session;
  }

  /**
   * Check if there's an active session
   */
  isActive(): boolean {
    return this.session !== null && !this.session.terminated;
  }

  /**
   * Parse violation data
   */
  private parseViolation(data: Record<string, any>): GuardrailViolation {
    const enforcement = data.enforcementLevel || data.enforcement_level || 'advisory';
    return {
      policyId: data.policyId || data.policy_id || '',
      policyName: data.policyName || data.policy_name || '',
      policyType: data.policyType || data.policy_type || '',
      violationType: data.violationType || data.violation_type || '',
      message: data.message || '',
      severity: data.severity || '',
      enforcementLevel: enforcement as EnforcementLevel,
      details: data.details,
    };
  }

  /**
   * Close the client and release resources
   */
  close(): void {
    if (this.wsClient) {
      this.wsClient.disconnect();
      this.wsClient = null;
    }
    this.session = null;
  }
}

/**
 * Wrap an async token stream with guardrail protection
 */
export async function* streamWithGuardrails(
  guardrail: StreamingGuardrail,
  tokenStream: AsyncIterable<string>,
  inputText?: string,
): AsyncGenerator<string> {
  await guardrail.startSession({ input: inputText });

  try {
    for await (const token of tokenStream) {
      for await (const filteredToken of guardrail.evaluate(token)) {
        yield filteredToken;
      }
    }
  } finally {
    if (guardrail.isActive()) {
      await guardrail.completeSession();
    }
  }
}

File: go/guardrails/streaming.go
// Package guardrails provides streaming guardrails for real-time LLM output validation.
package guardrails

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"
)

// StreamingGuardrail provides token-by-token evaluation of LLM output
// against configured guardrail policies with early termination support.
//
// Example:
//
//	config := StreamingGuardrailConfig{
//		APIKey:         "dx_...",
//		OrganizationID: "org_123",
//		ProjectID:      "proj_456",
//	}
//	guardrail := NewStreamingGuardrail(config)
//
//	session, err := guardrail.StartSession(ctx, nil)
//	if err != nil {
//		log.Fatal(err)
//	}
//
//	for token := range tokenStream {
//		filtered, err := guardrail.Evaluate(ctx, token, false)
//		if err != nil {
//			var violationErr *ViolationError
//			if errors.As(err, &violationErr) {
//				log.Printf("Blocked: %s", violationErr.Violation.Message)
//				break
//			}
//			log.Fatal(err)
//		}
//		fmt.Print(filtered)
//	}
type StreamingGuardrail struct {
	config     StreamingGuardrailConfig
	httpClient *http.Client
	session    *StreamingGuardrailSession
	tokenIndex int
	mu         sync.RWMutex
}

// StreamingGuardrailConfig holds configuration for StreamingGuardrail
type StreamingGuardrailConfig struct {
	APIKey                 string
	OrganizationID         string
	ProjectID              string
	BaseURL                string
	Timeout                time.Duration
	EvaluateEveryNTokens   int
	EnableEarlyTermination bool
	Debug                  bool
}

// StreamingGuardrailSession represents an active streaming session
type StreamingGuardrailSession struct {
	SessionID        string
	OrganizationID   string
	ProjectID        string
	ActivePolicies   []string
	TokensProcessed  int
	Violations       []Violation
	Terminated       bool
	TerminationReason string
	Allowed          bool
	AccumulatedText  string
}

// ViolationError is returned when a blocking guardrail violation occurs
type ViolationError struct {
	Violation Violation
	Session   *StreamingGuardrailSession
}

func (e *ViolationError) Error() string {
	return fmt.Sprintf("guardrail violation: %s", e.Violation.Message)
}

// EvaluateOptions contains options for token evaluation
type EvaluateOptions struct {
	TokenIndex *int
	IsLast     bool
}

// NewStreamingGuardrail creates a new streaming guardrail client
func NewStreamingGuardrail(config StreamingGuardrailConfig) *StreamingGuardrail {
	if config.BaseURL == "" {
		config.BaseURL = "https://api.diagnyx.io"
	}
	if config.Timeout == 0 {
		config.Timeout = 30 * time.Second
	}
	if config.EvaluateEveryNTokens == 0 {
		config.EvaluateEveryNTokens = 10
	}

	return &StreamingGuardrail{
		config: config,
		httpClient: &http.Client{
			Timeout: config.Timeout,
		},
	}
}

func (sg *StreamingGuardrail) log(msg string) {
	if sg.config.Debug {
		fmt.Printf("[DiagnyxGuardrails] %s\n", msg)
	}
}

func (sg *StreamingGuardrail) getBaseEndpoint() string {
	return fmt.Sprintf("%s/api/v1/organizations/%s/guardrails",
		strings.TrimSuffix(sg.config.BaseURL, "/"),
		sg.config.OrganizationID)
}

// StartSession starts a new streaming guardrail session
func (sg *StreamingGuardrail) StartSession(ctx context.Context, input *string) (*StreamingGuardrailSession, error) {
	sg.mu.Lock()
	defer sg.mu.Unlock()

	payload := map[string]interface{}{
		"projectId":              sg.config.ProjectID,
		"evaluateEveryNTokens":   sg.config.EvaluateEveryNTokens,
		"enableEarlyTermination": sg.config.EnableEarlyTermination,
	}
	if input != nil {
		payload["input"] = *input
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		sg.getBaseEndpoint()+"/evaluate/stream/start", bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+sg.config.APIKey)
	req.Header.Set("Accept", "application/json")

	resp, err := sg.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	var data map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	eventType, _ := data["type"].(string)
	if eventType == "session_started" {
		sessionID, _ := data["sessionId"].(string)
		policies := getStringSlice(data, "activePolicies")

		sg.session = &StreamingGuardrailSession{
			SessionID:      sessionID,
			OrganizationID: sg.config.OrganizationID,
			ProjectID:      sg.config.ProjectID,
			ActivePolicies: policies,
			Allowed:        true,
		}
		sg.tokenIndex = 0
		sg.log(fmt.Sprintf("Session started: %s", sessionID))
		return sg.session, nil
	} else if eventType == "error" {
		errorMsg, _ := data["error"].(string)
		return nil, fmt.Errorf("failed to start session: %s", errorMsg)
	}

	return nil, errors.New("unexpected response type")
}

// Evaluate evaluates a token against guardrail policies
// Returns the token if it passes validation, empty string if blocked,
// and an error if a blocking violation occurred.
func (sg *StreamingGuardrail) Evaluate(ctx context.Context, token string, isLast bool) (string, error) {
	return sg.EvaluateWithOptions(ctx, token, EvaluateOptions{IsLast: isLast})
}

// EvaluateWithOptions evaluates a token with additional options
func (sg *StreamingGuardrail) EvaluateWithOptions(ctx context.Context, token string, opts EvaluateOptions) (string, error) {
	sg.mu.Lock()
	defer sg.mu.Unlock()

	if sg.session == nil {
		return "", errors.New("no active session, call StartSession first")
	}

	tokenIndex := sg.tokenIndex
	if opts.TokenIndex != nil {
		tokenIndex = *opts.TokenIndex
	}
	sg.tokenIndex++

	sg.session.AccumulatedText += token

	payload := map[string]interface{}{
		"sessionId":  sg.session.SessionID,
		"token":      token,
		"tokenIndex": tokenIndex,
		"isLast":     opts.IsLast,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		sg.getBaseEndpoint()+"/evaluate/stream", bytes.NewReader(body))
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+sg.config.APIKey)
	req.Header.Set("Accept", "text/event-stream")

	resp, err := sg.httpClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	reader := bufio.NewReader(resp.Body)
	var result string

	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			if err == io.EOF {
				break
			}
			return result, fmt.Errorf("error reading stream: %w", err)
		}

		line = strings.TrimSpace(line)
		if !strings.HasPrefix(line, "data: ") {
			continue
		}

		jsonData := line[6:]
		var data map[string]interface{}
		if err := json.Unmarshal([]byte(jsonData), &data); err != nil {
			sg.log(fmt.Sprintf("Failed to parse event: %v", err))
			continue
		}

		eventType, _ := data["type"].(string)

		switch eventType {
		case "token_allowed":
			idx, _ := data["tokenIndex"].(float64)
			sg.session.TokensProcessed = int(idx) + 1
			result = token

		case "violation_detected":
			violation := sg.parseViolation(data)
			sg.session.Violations = append(sg.session.Violations, violation)
			if violation.EnforcementLevel == EnforcementBlocking {
				sg.session.Allowed = false
			}

		case "early_termination":
			blockingData, _ := data["blockingViolation"].(map[string]interface{})
			violation := sg.parseViolation(blockingData)
			sg.session.Terminated = true
			reason, _ := data["reason"].(string)
			sg.session.TerminationReason = reason
			sg.session.Allowed = false
			return "", &ViolationError{
				Violation: violation,
				Session:   sg.session,
			}

		case "session_complete":
			totalTokens, _ := data["totalTokens"].(float64)
			allowed, _ := data["allowed"].(bool)
			sg.session.TokensProcessed = int(totalTokens)
			sg.session.Allowed = allowed

		case "error":
			errorMsg, _ := data["error"].(string)
			sg.log(fmt.Sprintf("Error: %s", errorMsg))
		}
	}

	return result, nil
}

// EvaluateChannel evaluates tokens from a channel and sends results to output channel
func (sg *StreamingGuardrail) EvaluateChannel(ctx context.Context, tokens <-chan string, markLast func(string) bool) (<-chan string, <-chan error) {
	results := make(chan string, 10)
	errors := make(chan error, 1)

	go func() {
		defer close(results)
		defer close(errors)

		for token := range tokens {
			select {
			case <-ctx.Done():
				errors <- ctx.Err()
				return
			default:
				isLast := markLast != nil && markLast(token)
				result, err := sg.Evaluate(ctx, token, isLast)
				if err != nil {
					errors <- err
					return
				}
				if result != "" {
					results <- result
				}
			}
		}
	}()

	return results, errors
}

// CompleteSession completes the current session
func (sg *StreamingGuardrail) CompleteSession(ctx context.Context) (*StreamingGuardrailSession, error) {
	sg.mu.Lock()
	defer sg.mu.Unlock()

	if sg.session == nil {
		return nil, errors.New("no active session")
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		fmt.Sprintf("%s/evaluate/stream/%s/complete", sg.getBaseEndpoint(), sg.session.SessionID), nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+sg.config.APIKey)
	req.Header.Set("Accept", "text/event-stream")

	resp, err := sg.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	reader := bufio.NewReader(resp.Body)
	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			if err == io.EOF {
				break
			}
			break
		}

		line = strings.TrimSpace(line)
		if !strings.HasPrefix(line, "data: ") {
			continue
		}

		var data map[string]interface{}
		if err := json.Unmarshal([]byte(line[6:]), &data); err != nil {
			continue
		}

		if data["type"] == "session_complete" {
			totalTokens, _ := data["totalTokens"].(float64)
			allowed, _ := data["allowed"].(bool)
			sg.session.TokensProcessed = int(totalTokens)
			sg.session.Allowed = allowed
		}
	}

	session := sg.session
	sg.session = nil
	return session, nil
}

// CancelSession cancels the current session
func (sg *StreamingGuardrail) CancelSession(ctx context.Context) (bool, error) {
	sg.mu.Lock()
	defer sg.mu.Unlock()

	if sg.session == nil {
		return false, nil
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodDelete,
		fmt.Sprintf("%s/evaluate/stream/%s", sg.getBaseEndpoint(), sg.session.SessionID), nil)
	if err != nil {
		return false, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+sg.config.APIKey)

	resp, err := sg.httpClient.Do(req)
	if err != nil {
		return false, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return false, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	var result struct {
		Cancelled bool `json:"cancelled"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return false, fmt.Errorf("failed to decode response: %w", err)
	}

	sg.session = nil
	return result.Cancelled, nil
}

// GetSession returns the current session
func (sg *StreamingGuardrail) GetSession() *StreamingGuardrailSession {
	sg.mu.RLock()
	defer sg.mu.RUnlock()
	return sg.session
}

// IsActive returns true if there's an active session
func (sg *StreamingGuardrail) IsActive() bool {
	sg.mu.RLock()
	defer sg.mu.RUnlock()
	return sg.session != nil && !sg.session.Terminated
}

func (sg *StreamingGuardrail) parseViolation(data map[string]interface{}) Violation {
	if data == nil {
		return Violation{}
	}

	enforcement := getString(data, "enforcementLevel", "enforcement_level")
	level := EnforcementAdvisory
	if enforcement != "" {
		level = EnforcementLevel(enforcement)
	}

	return Violation{
		PolicyID:         getString(data, "policyId", "policy_id"),
		PolicyName:       getString(data, "policyName", "policy_name"),
		PolicyType:       getString(data, "policyType", "policy_type"),
		ViolationType:    getString(data, "violationType", "violation_type"),
		Message:          getString(data, "message"),
		Severity:         getString(data, "severity"),
		EnforcementLevel: level,
		Details:          getMap(data, "details"),
	}
}

// StreamWithGuardrails wraps a token channel with guardrail protection
func StreamWithGuardrails(
	ctx context.Context,
	config StreamingGuardrailConfig,
	tokens <-chan string,
	input *string,
	markLast func(string) bool,
) (<-chan string, <-chan error) {
	results := make(chan string, 10)
	errors := make(chan error, 1)

	go func() {
		defer close(results)
		defer close(errors)

		guardrail := NewStreamingGuardrail(config)

		_, err := guardrail.StartSession(ctx, input)
		if err != nil {
			errors <- err
			return
		}

		for token := range tokens {
			select {
			case <-ctx.Done():
				errors <- ctx.Err()
				return
			default:
				isLast := markLast != nil && markLast(token)
				result, err := guardrail.Evaluate(ctx, token, isLast)
				if err != nil {
					errors <- err
					return
				}
				if result != "" {
					results <- result
				}
			}
		}

		if guardrail.IsActive() {
			_, _ = guardrail.CompleteSession(ctx)
		}
	}()

	return results, errors
}
